apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/RpcRetryConsumer.cs

```
@@ -0,0 +1,214 @@
+        private int _retries = 0;        
+
+        internal RpcRetryConsumer(IConsumer<T> consumer, RetryOptions options, 
+            Func<T, int> sizer, Func<Action, RetryOptions, ISequentialThreadingTimer> timerFactory)
+        {
+            _consumer = GaxPreconditions.CheckNotNull(consumer, nameof(consumer));
+            _options = GaxPreconditions.CheckNotNull(options, nameof(options));
+            _sizer = GaxPreconditions.CheckNotNull(sizer, nameof(sizer));
+            _timerFactory = GaxPreconditions.CheckNotNull(timerFactory, nameof(timerFactory));
+
+            // Start the background timer that will retry failed requests.
+            _timer = _timerFactory(FlushBuffer, _options);
+        }
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _consumer.Receive(items);
+            }
+            catch (RpcException e)
+            {
+                HandleException(items, e);
+            }
+        }
+
+        /// <inheritdoc />
+        public Task ReceiveAsync(IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                return _consumer.ReceiveAsync(items, cancellationToken);
+            }
+            catch (RpcException e)
+            {
+                HandleException(items, e);
+                return CommonUtils.CompletedTask;
+            }
+        }
+
+        /// <inheritdoc />
+        public void Dispose() => _timer?.Dispose();
+
+        /// <summary>
+        /// Handles a <see cref="RpcException"/> that occurred during a <see cref="Receive(IEnumerable{T})"/> call.
+        /// </summary>
+        private void HandleException(IEnumerable<T> items, RpcException e)
```

Maybe have this return a bool and use it in an exception filter where this is needed. Return false when you want to propagate the exception so you don't lose stack trace details.